### PR TITLE
planner: add trace for join reorder

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/29661

### What is changed and how it works?

add trace for join reorder.

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/rule_join_reorder_greedy.go [hunk 1]
```diff
@@ -101,6 +103,7 @@ func (s *joinReorderGreedySolver) constructConnectedJoinTree() (*jrNode, error)
 			p:       bestJoin,
 			cumCost: bestCost,
 		}
+		tracer.appendLogicalJoinCost(bestJoin, bestCost)
```
Comment (time-and-fate) [inline planner/core/rule_join_reorder_greedy.go:106]: Can we have test cases to cover the cost of joins traced here?

File: planner/core/rule_join_reorder.go [hunk 2]
```diff
@@ -194,3 +204,116 @@ func (s *baseSingleGroupJoinOrderSolver) calcJoinCumCost(join LogicalPlan, lNode
...
 }
+
+func appendJoinReorderTraceStep(tracer *joinReorderTrace, plan LogicalPlan, opt *logicalOptimizeOp) {
+	if len(tracer.initial) < 1 || len(tracer.final) < 1 {
+		return
+	}
+	if tracer.final != tracer.initial {
+		action := fmt.Sprintf("join order become %v from origin %v", tracer.final, tracer.initial)
+		reason := func() string {
+			buffer := bytes.NewBufferString("join cost during reorder[")
```
Comment (time-and-fate) [inline planner/core/rule_join_reorder.go:215]:
```suggestion
			buffer := bytes.NewBufferString("join cost during reorder: [")
```

File: planner/core/rule_join_reorder.go [hunk 3]
```diff
@@ -194,3 +204,116 @@ func (s *baseSingleGroupJoinOrderSolver) calcJoinCumCost(join LogicalPlan, lNode
 func (*joinReOrderSolver) name() string {
 	return "join_reorder"
 }
+
+func appendJoinReorderTraceStep(tracer *joinReorderTrace, plan LogicalPlan, opt *logicalOptimizeOp) {
+	if len(tracer.initial) < 1 || len(tracer.final) < 1 {
+		return
+	}
+	if tracer.final != tracer.initial {
+		action := fmt.Sprintf("join order become %v from origin %v", tracer.final, tracer.initial)
```
Comment (time-and-fate) [inline planner/core/rule_join_reorder.go:213]:
```suggestion
		action := fmt.Sprintf("join order becomes %v from original %v", tracer.final, tracer.initial)
```

File: planner/core/rule_join_reorder_greedy.go (hunk 1 above)
Comment (Yisaer) [reply planner/core/rule_join_reorder_greedy.go:106]: updated

**Review by time-and-fate — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]

**Pull request merged**